### PR TITLE
Adding an ability to set a certain check_type when setting up MySQL replication host groups

### PR DESCRIPTION
**Pull request by zentavr**
##### SUMMARY
Adds a feature described at #19

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
proxysql_replication_hostgroup

### CODE REVIEW

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 1]
```diff
@@ -33,6 +33,12 @@
     description:
       - Text field that can be used for any purposes defined by the user.
     type: str
+  check_type:
+    description:
+      - Which check type to use when detecting that the node is a slave.
+    type: str
+    choices: [ "read_only", "innodb_read_only", "super_read_only", "read_only|innodb_read_only", "read_only&innodb_read_only" ]
+    default: read_only
```
Comment (felixfontein) [inline plugins/modules/proxysql_replication_hostgroups.py:41]:
```suggestion
    default: read_only
    version_added: 1.1.0
```

**Review by felixfontein — COMMENTED** [review]
I don't know proxysql so I cannot say much about your change except some formal stuff :) You definitely need a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 2]
```diff
@@ -33,6 +33,12 @@
     description:
       - Text field that can be used for any purposes defined by the user.
     type: str
+  check_type:
+    description:
+      - Which check type to use when detecting that the node is a slave.
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:38]:
```suggestion
      - Which check type to use when detecting that the node is a standby.
```
just a suggestion. or we could use `replica` alternatively

**Review by Andersson007 — COMMENTED** [review]
@zentavr thanks for the contribution!

In addition to what @felixfontein mentioned, could you please cover the option in CI tests, see https://github.com/ansible-collections/community.proxysql/tree/main/tests/integration/targets

File: plugins/modules/proxysql_replication_hostgroups.py (hunk 1 above)
Comment (zentavr) [reply plugins/modules/proxysql_replication_hostgroups.py:41]: Added as 725e02d0ba0779861df45914f9e83f97df5ed4d9

File: plugins/modules/proxysql_replication_hostgroups.py (hunk 2 above)
Comment (zentavr) [reply plugins/modules/proxysql_replication_hostgroups.py]: Done in 936071b106cbe536211938bec813bc0d8e89b703

File: plugins/modules/proxysql_replication_hostgroups.py (hunk 1 above)
Comment (zentavr) [reply plugins/modules/proxysql_replication_hostgroups.py:41]: @felixfontein added a changelog in be9aa752b4d14421fddee19e0420ad8d58f019ac

File: plugins/modules/proxysql_replication_hostgroups.py (hunk 2 above)
Comment (zentavr) [reply plugins/modules/proxysql_replication_hostgroups.py]: As for the tests - will do that later.

**Comment by felixfontein** [comment]
@zentavr could you please remove the `@` mentions from the commit messages? Mentioning people in commit messages causes a flood of notifications for the mentioned persons (every time this commit is pushed to some repo), which can get pretty annoying :)

**Comment by zentavr** [comment]
@felixfontein - I had removed the mentions.

**Comment by felixfontein** [comment]
@zentavr thanks! :)

**Comment by zentavr** [comment]
@Andersson007 - added CI tests

**Comment by Andersson007** [comment]
@zentavr thanks for the tests!
I'm not a proxysql user, so can't judge anything except general things like doc formatting and tests.
We should get feedback from users.
Also a couple of general questions:
1. is this feature common for all proxysql versions? I mean, if it's been introduced recently, for example, won't the changes somehow break current users playbooks using the older proxysql versions / driver versions?
2. is it possible to add a case with `check_mode: yes` to the tests?

**Comment by zentavr** [comment]
@Andersson007 [here](http://www.mafiree.com/blogs.php?ref=Whats-new-in-ProxySQL-2.0) I had found that `check_type` appeared since 2.x version of ProxySQL. Saying, the release date of 2.0.2 is 2019-02-18 (this is what I had found in [ProxySQL Repo](https://github.com/sysown/proxysql/)).

As for your second question: As I understood you correctly, you do want to have tests with `check_type` option and without it - correct?

**Comment by zentavr** [comment]
The tests of the module rely on 2.0.12 version of ProxySQL which has `check_type`. At the early beginning there were only 3 possible values to specify for the `check_type`.

**Comment by Andersson007** [comment]
> As for your second question: As I understood you correctly, you do want to have tests with `check_type` option and without it - correct?

Not really, i meant running in check mode, i.e., passing `check_mode: yes` as an option (because the module supports the check mode, so better to cover this case too)

> @Andersson007 [here](http://www.mafiree.com/blogs.php?ref=Whats-new-in-ProxySQL-2.0) I had found that `check_type` appeared since 2.x version of ProxySQL. Saying, the release date of 2.0.2 is 2019-02-18 (this is what I had found in [ProxySQL Repo](https://github.com/sysown/proxysql/)).

So, will these changes break users playbooks if they use ProxySQL versions of lower than 2.0.2?
If yes:
1. We should mention in the option's description that this feature is supported since ProxySQL 2.0.2, otherwise ignored
2. Get and check the version in the code: if 2.0.2 and higher, append the option, else, pass. You can use this PR https://github.com/ansible-collections/community.mysql/pull/71 as an example (changes in mysql_replication.py and the unit tests)

**Comment by Andersson007** [comment]
any updates on this?

**Comment by zentavr** [comment]
@Andersson007 Currently very busy with my regular work, so I had no chance to check up your comments.

**Comment by Andersson007** [comment]
@zentavr if you're busy what about if anyone else complete it, what do you think?
@bmildren as the modules author could you please take a look?

**Comment by zentavr** [comment]
@Andersson007 I’m ok with your proposal.
Seems like the tests is the only open question

File: changelogs/changelog.yaml [hunk 3]
```diff
@@ -1,5 +1,11 @@
 ancestor: null
 releases:
+  1.1.0:
+    changes:
+      minor_changes:
+        - Adds an ability to specify which ProxySQL check type to use when detecting
+          a MySQL standby node in proxysql_replication_hostgroups module.
```
Comment (felixfontein) [inline changelogs/changelog.yaml:7]: Please do NOT modify `changelogs/changelog.yaml` directly. Create a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

**Review by felixfontein — CHANGES_REQUESTED** [review]

**Comment by Andersson007** [comment]
@zentavr as the only contributor to the collection so far, do you have any ideas if the PR https://github.com/ansible-collections/community.proxysql/pull/25 brings value? Maybe you work with the module.

**Comment by zentavr** [comment]
Hello @Andersson007. We do not use query_rules so heavy (we've implemented ProxySQL few month ago into our infrastructure), but the proposed features sound very interesting.

**Comment by Andersson007** [comment]
Hello @zentavr , cool:) Thanks for the feedback!

**Comment by markuman** [comment]
I'll take a look.  
[In the past](https://github.com/ansible/ansible/pull/60268) I was failing to implement it, because of breaking backwards compatibility and the fast-moving and chaning proxysql itself.

**Comment by markuman** [comment]
Yeah, it's the same issue. The `check_type` schema was introduced with 2.0.1.  
All users that are using 2.0.0 or the 1.4.x releases won't be able to use this module anymore.

**Comment by zentavr** [comment]
@markuman actually if they won't specify the check_type - nothing should happen.

**Comment by markuman** [comment]
> @markuman actually if they won't specify the check_type - nothing should happen.

It don't looks like, [because `check_type` comes with a default value `read_only`](https://github.com/ansible-collections/community.proxysql/pull/20/files#diff-69718fdd40ccd856ae2327f41549f205249e8d7f142d72494a54c8f891a70eacR331) and the queries are all extended with that column.

**Comment by zentavr** [comment]
@markuman I see. What could be the workaround?

from one side we could try to stick to the very default functionality and be compatible with old and new versions of ProxySQL.

another solution could be a ProxySQL version checking before we start any SQL questions and being silent or throwing the exception of the user specified the parameter which is not supported by the ProxySQL version.

**Comment by Andersson007** [comment]
Checking version sounds like a good solution. We do it in community.mysql:
```
./plugins/module_utils/mysql.py get_server_version function
./plugins/module_utils/implementations/mysql/replication.py 
./plugins/module_utils/implementations/mysql/user.py
```

**Comment by markuman** [comment]
> @markuman I see. What could be the workaround?

@zentavr 

modify the sql query for every version results in many if/else statements.The code gets confusing. That was what I was trying in the past.  

From today's perspective I would try to leave the "_very default functionality_" as it is.  
And when `check_type` is requested and the underlying proxysql version supports it, I would introduce `update` statements, that'll handle only the `check_type` column - in a new function that is separated from the "_very default functionality_".  
That keeps the code clean, maintanable and can be a path for further proxysql version-based features/parameters and help to keep backward compatibility.

**Comment by markuman** [comment]
Implemented via https://github.com/ansible-collections/community.proxysql/pull/69

**Comment by Andersson007** [comment]
thanks you all folks!

**Pull request closed without merging**